File: tools/linkedin-scrape/li-scrape.js
#!/usr/bin/env node
/**
 * li-scrape.js — free LinkedIn profile -> analyzable data. No paid API, no OCR.
 *
 * Drives an already-authenticated Chrome over CDP, visits each profile section,
 * and extracts the RENDERED TEXT from the DOM (exact — not screenshots, not OCR).
 * Screenshots are saved alongside only as a human-viewable artifact / vision
 * fallback. A best-effort structured parse of the rendered text is written too;
 * for a rich parse, hand `profile.raw.txt` to an LLM.
 *
 * The pure text/URL helpers are exported for unit testing; the browser runner
 * only executes when this file is run directly (see the bottom of the file).
 *
 * One-time setup (headed Chrome you log into once):
 *   "$CHROME_BIN" --user-data-dir="$HOME/.cache/pw-li-profile" \
 *       --remote-debugging-port=9777 --no-first-run about:blank &
 *   # ...log into LinkedIn once; the session persists in the profile dir.
 *
 * Usage:
 *   CHROME_BIN=/path/to/chrome node li-scrape.js <profile-url-or-public-id> [outDir]
 */
'use strict';

const CDP_ENDPOINT = 'http://127.0.0.1:9777';

/** Profile sections to visit. '' is the top card + activity. */
const SECTIONS = ['', 'details/experience', 'details/education', 'details/skills', 'details/interests'];

/** Lines that mark the end of a profile's own content on a details page. */
const SECTION_END_MARKERS = [
  'More profiles for you',
  'People you may know',
  'You might like',
  'Explore Premium profiles',
];

// ---------------------------------------------------------------------------
// Pure helpers (exported, unit-tested)
// ---------------------------------------------------------------------------

/**
 * Extract the public id from a profile URL or a bare id.
 * Accepts full URLs (any section/query/trailing slash), `/in/<id>` fragments,
 * or the bare id itself. Returns null if nothing usable is found.
 */
function parsePublicId(input) {
  if (typeof input !== 'string') return null;
  let s = input.trim();
  if (!s) return null;
  s = s.split(/[?#]/)[0];           // drop query/hash
  s = s.replace(/\/+$/, '');        // drop trailing slashes
  if (s.includes('/in/')) s = s.split('/in/')[1];
  else if (s.includes('linkedin.com')) return null; // a LinkedIn URL but not a profile
  const id = s.split('/')[0].trim();
  return id || null;
}

/** Canonical profile base URL for a public id. */
function profileBase(publicId) {
  return `https://www.linkedin.com/in/${publicId}`;
}

/** URL for a given section ('' = top card). */
function sectionUrl(publicId, section) {
  const base = profileBase(publicId);
  return section ? `${base}/${section}/` : `${base}/`;
}

/** Short filesystem-safe name for a section ('' -> 'profile'). */
function sectionName(section) {
  return section ? section.split('/').pop() : 'profile';
}

/** True if a URL indicates we are logged out / gated. */
function isBlockedUrl(url) {
  return typeof url === 'string' && /(authwall|\/login|\/checkpoint\/|\/uas\/login)/.test(url);
}

/** Resolve the output directory for a run. */
function outDirFor(cwd, publicId, override) {
  const path = require('path');
  return override ? path.resolve(cwd, override) : path.join(cwd, `out-${publicId}`);
}

/**
 * Split raw innerText into trimmed non-empty lines, collapsing runs of the same
 * line (LinkedIn renders a visible + a visually-hidden copy of most text).
 */
function normalizeLines(raw) {
  const out = [];
  for (const line of String(raw || '').split('\n')) {
    const t = line.trim();
    if (!t) continue;
    if (out[out.length - 1] !== t) out.push(t);
  }
  return out;
}

/**
 * Return the lines of one section: from the line equal to `heading` up to the
 * first end-marker (exclusive). If the heading is absent, returns all lines
 * (already normalized), so callers still get something to work with.
 */
function sliceSection(raw, heading, endMarkers = SECTION_END_MARKERS) {
  const lines = normalizeLines(raw);
  let start = lines.findIndex((l) => l === heading);
  start = start === -1 ? 0 : start + 1;
  let end = lines.length;
  for (let i = start; i < lines.length; i++) {
    if (endMarkers.some((m) => lines[i] === m || lines[i].startsWith(m))) { end = i; break; }
  }
  return lines.slice(start, end);
}

/** Parse a "Company · Full-time" style line into {company, employmentType}. */
function splitCompanyType(line) {
  const parts = String(line).split('·').map((s) => s.trim());
  return { company: parts[0] || null, employmentType: parts[1] || null };
}

/** Parse a "Location · Remote" style line into {location, arrangement}. */
function splitLocationArrangement(line) {
  const parts = String(line).split('·').map((s) => s.trim());
  const arr = parts.length > 1 ? parts[parts.length - 1] : null;
  const known = /^(Remote|Hybrid|On-site|Onsite)$/i;
  if (arr && known.test(arr)) {
    return { location: parts.slice(0, -1).join(' · ') || null, arrangement: arr };
  }
  return { location: line.trim() || null, arrangement: null };
}

const DATE_LINE = /(present|\d{4}|jan|feb|mar|apr|may|jun|jul|aug|sep|oct|nov|dec)/i;
const DURATION_TAIL = /·\s*[^·]*\b(yr|yrs|mo|mos)\b[^·]*$/i;

/** True if a line looks like a LinkedIn date range ("Apr 2026 - Present · 5 mos"). */
function looksLikeDateLine(line) {
  return DATE_LINE.test(line) && /[-–—]|present/i.test(line);
}

/** Strip the "· 5 mos" duration tail from a date line. */
function stripDuration(line) {
  return String(line).replace(DURATION_TAIL, '').trim();
}

/**
 * Best-effort parse of the Experience details text into position records.
 * Heuristic: an entry is a title line immediately followed by a "Company · Type"
 * line; the next date line and location line are attached; remaining lines up to
 * the next entry (or a "…skills" trailer) become the description.
 */
function parseExperience(raw) {
  const lines = sliceSection(raw, 'Experience');
  const positions = [];
  for (let i = 0; i < lines.length; i++) {
    const next = lines[i + 1];
    if (next && next.includes('·') && !looksLikeDateLine(next) && !looksLikeDateLine(lines[i])) {
      const { company, employmentType } = splitCompanyType(next);
      const pos = { title: lines[i], company, employmentType, dateRange: null, location: null, arrangement: null, description: [] };
      let j = i + 2;
      if (lines[j] && looksLikeDateLine(lines[j])) { pos.dateRange = stripDuration(lines[j]); j++; }
      if (lines[j] && !looksLikeDateLine(lines[j]) && !(lines[j + 1] && lines[j + 1].includes('·') && !looksLikeDateLine(lines[j + 1]))) {
        const { location, arrangement } = splitLocationArrangement(lines[j]);
        pos.location = location; pos.arrangement = arrangement; j++;
      }
      while (j < lines.length) {
        const l = lines[j];
        const peek = lines[j + 1];
        if (peek && peek.includes('·') && !looksLikeDateLine(peek)) break; // next entry
        if (/\band \+\d+ skills?$/.test(l) || /^…?\s*more$/i.test(l) || l === 'Show translation') { j++; continue; }
        pos.description.push(l);
        j++;
      }
      positions.push(pos);
      i = j - 1;
    }
  }
  return positions;
}

/**
 * Best-effort parse of Education details text.
 * Entry shape: school / "Degree, Field" / date range.
 */
function parseEducation(raw) {
  const lines = sliceSection(raw, 'Education');
  const edu = [];
  for (let i = 0; i < lines.length; i++) {
    const detail = lines[i + 1];
    const date = lines[i + 2];
    if (detail && date && looksLikeDateLine(date) && !looksLikeDateLine(lines[i]) && !looksLikeDateLine(detail)) {
      const [degree, ...rest] = detail.split(',');
      edu.push({
        school: lines[i],
        degree: rest.length ? degree.trim() : null,
        field: rest.length ? rest.join(',').trim() : degree.trim(),
        dateRange: stripDuration(date),
      });
      i += 2;
    }
  }
  return edu;
}

/**
 * Best-effort parse of Skills details text -> [{name, endorsements}].
 * A skill is a line whose following lines may include "N endorsement(s)".
 * Lines that are sub-labels ("... at Company", "Endorsed") are skipped.
 */
function parseSkills(raw) {
  const lines = sliceSection(raw, 'Skills').filter((l) => l !== 'All');
  const skills = [];
  const isNoise = (l) => /\bendorsement/i.test(l) || l === 'Endorsed' || / at /.test(l) || /^Show/.test(l);
  for (let i = 0; i < lines.length; i++) {
    const l = lines[i];
    if (isNoise(l)) continue;
    // look ahead for an endorsement count within the next 3 lines
    let endorsements = 0;
    for (let k = 1; k <= 3 && lines[i + k]; k++) {
      const m = lines[i + k].match(/(\d+)\s+endorsement/i);
      if (m) { endorsements = Number(m[1]); break; }
    }
    if (!skills.some((s) => s.name === l)) skills.push({ name: l, endorsements });
  }
  return skills;
}

/** Pull the first email address out of contact-modal text, or null. */
function extractEmail(text) {
  const m = String(text || '').match(/[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,}/i);
  return m ? m[0] : null;
}

const helpers = {
  CDP_ENDPOINT, SECTIONS, SECTION_END_MARKERS,
  parsePublicId, profileBase, sectionUrl, sectionName, isBlockedUrl, outDirFor,
  normalizeLines, sliceSection, splitCompanyType, splitLocationArrangement,
  looksLikeDateLine, stripDuration, parseExperience, parseEducation, parseSkills,
  extractEmail,
};

module.exports = helpers;

// ---------------------------------------------------------------------------
// Browser runner (only when executed directly)
// ---------------------------------------------------------------------------

async function run() {
  const fs = require('fs');
  const path = require('path');
  const { chromium } = require('playwright-core');

  const arg = process.argv[2];
  if (!arg) { console.error('usage: node li-scrape.js <profile-url-or-public-id> [outDir]'); process.exit(1); }
  const publicId = parsePublicId(arg);
  if (!publicId) { console.error(`could not parse a public id from: ${arg}`); process.exit(1); }

  const outDir = outDirFor(process.cwd(), publicId, process.argv[3]);
  const rawDir = path.join(outDir, 'raw');
  fs.mkdirSync(rawDir, { recursive: true });

  const browser = await chromium.connectOverCDP(CDP_ENDPOINT).catch(() => {
    console.error(`Cannot reach Chrome at ${CDP_ENDPOINT}. Start it and log into LinkedIn once (see header).`);
    process.exit(1);
  });
  const ctx = browser.contexts()[0];
  const page = ctx.pages().find((p) => p.url().includes('linkedin.com')) || (await ctx.newPage());
  await page.setViewportSize({ width: 1440, height: 1000 });

  const sleep = (ms) => new Promise((r) => setTimeout(r, ms));
  const texts = {};

  for (const section of SECTIONS) {
    const name = sectionName(section);
    process.stderr.write(`  ${name} … `);
    await page.goto(sectionUrl(publicId, section), { waitUntil: 'domcontentloaded', timeout: 60000 });
    await sleep(3500);
    if (isBlockedUrl(page.url())) { console.error('\nNOT LOGGED IN — log into LinkedIn in the Chrome window first.'); process.exit(2); }
    await page.evaluate(async () => {
      const a = document.querySelector('aside.msg-overlay-list-bubble, aside#msg-overlay'); if (a) a.style.display = 'none';
      document.querySelectorAll('main button').forEach((b) => { if ((b.innerText || '').trim().toLowerCase().includes('see more')) b.click(); });
      const s = document.scrollingElement;
      for (let y = 0; y < s.scrollHeight; y += 700) { s.scrollTop = y; await new Promise((r) => setTimeout(r, 150)); }
      s.scrollTop = 0;
    });
    await sleep(800);
    const txt = await page.evaluate(() => (document.querySelector('main') || document.body).innerText);
    texts[name] = txt;
    fs.writeFileSync(path.join(rawDir, `${name}.txt`), txt);
    await page.screenshot({ path: path.join(rawDir, `${name}.png`), fullPage: true }).catch(() => {});
    console.error(`${txt.length} chars`);
  }

  // Contact-info modal (email lives here, not on the section pages).
  let contact = { email: null, raw: null };
  try {
    process.stderr.write('  contact … ');
    await page.goto(profileBase(publicId) + '/', { waitUntil: 'domcontentloaded', timeout: 60000 });
    await sleep(3000);
    await page.evaluate(() => {
      for (const el of document.querySelectorAll('a,button,span')) {
        const own = [...el.childNodes].filter((n) => n.nodeType === 3).map((n) => n.textContent.trim()).join('');
        if (own === 'Contact info') { el.click(); return; }
      }
    });
    await sleep(2000);
    const modal = await page.evaluate(() => {
      const m = document.querySelector('[role="dialog"], .artdeco-modal');
      return m ? m.innerText : null;
    });
    contact = { email: extractEmail(modal), raw: modal };
    if (modal) fs.writeFileSync(path.join(rawDir, 'contact.txt'), modal);
    await page.keyboard.press('Escape').catch(() => {});
    console.error(contact.email || '(none)');
  } catch (e) { console.error('(skipped:', e.message + ')'); }

  // Concatenated raw text for the LLM parse pass.
  fs.writeFileSync(path.join(outDir, 'profile.raw.txt'),
    Object.entries(texts).map(([k, v]) => `\n===== ${k} =====\n${v}`).join('\n'));

  // Best-effort structured parse.
  const structured = {
    publicId,
    source: profileBase(publicId),
    contact: { email: contact.email },
    positions: parseExperience(texts.experience || ''),
    education: parseEducation(texts.education || ''),
    skills: parseSkills(texts.skills || ''),
  };
  fs.writeFileSync(path.join(outDir, 'profile.json'), JSON.stringify(structured, null, 2));

  console.error(`\nDone -> ${outDir}`);
  console.error('  raw/*.txt + raw/*.png per section, contact.txt');
  console.error('  profile.json      (best-effort structured parse)');
  console.error('  profile.raw.txt   (feed to an LLM for a rich parse)');
  await browser.close(); // detaches CDP only; Chrome keeps running
}

if (require.main === module) {
  run().catch((e) => { console.error('ERR:', e.message); process.exit(1); });
}

File: tools/linkedin-scrape/li_scrape.py
# /// script
# requires-python = ">=3.11"
# dependencies = ["playwright>=1.40"]
# ///
"""li_scrape.py - free LinkedIn profile/job-post -> analyzable data. No paid API, no OCR.

Drives an already-authenticated Chrome over CDP, visits each profile section
(or a job post), and extracts the RENDERED TEXT from the DOM (exact - not
screenshots, not OCR). Screenshots are saved alongside only as a human-viewable
artifact / vision fallback. A best-effort structured parse is written too; for
a rich parse, hand `profile.raw.txt` to an LLM.

The pure text/URL helpers are plain functions covered by test_li_scrape.py;
the browser runner only executes under __main__.

One-time setup (headed Chrome you log into once):
    "$CHROME_BIN" --user-data-dir="$HOME/.cache/pw-li-profile" \
        --remote-debugging-port=9777 --no-first-run about:blank &
    # ...log into LinkedIn once; the session persists in the profile dir.

Usage:
    uv run li_scrape.py <profile-url-or-public-id> [outDir]
    uv run li_scrape.py <jobs-view-url-or-job-id>  [outDir]
"""
import json
import re
import sys
import time
from pathlib import Path

CDP_ENDPOINT = "http://127.0.0.1:9777"
SCHEMA_PROFILE = "li-scrape/1"
SCHEMA_JOB = "li-scrape-job/1"

# Profile sections to visit. '' is the top card + activity.
SECTIONS = ["", "details/experience", "details/education", "details/skills", "details/interests"]

# Lines that mark the end of a profile's own content on a details page.
SECTION_END_MARKERS = [
    "More profiles for you",
    "People you may know",
    "You might like",
    "Explore Premium profiles",
]

# Lines that mark the end of the job description block.
JOB_END_MARKERS = [
    "About the company",
    "Set alert for similar jobs",
    "How you match",
    "Benefits found from the job description",
]

# ---------------------------------------------------------------------------
# Pure helpers (unit-tested)
# ---------------------------------------------------------------------------

def parse_public_id(value):
    """Public id from a profile URL, an `/in/<id>` fragment, or the bare id."""
    if not isinstance(value, str):
        return None
    s = value.strip()
    if not s:
        return None
    s = re.split(r"[?#]", s)[0]
    s = s.rstrip("/")
    if "/in/" in s:
        s = s.split("/in/")[1]
    elif "linkedin.com" in s:
        return None  # a LinkedIn URL but not a profile
    return s.split("/")[0].strip() or None


def profile_base(public_id):
    return f"https://www.linkedin.com/in/{public_id}"


def section_url(public_id, section):
    base = profile_base(public_id)
    return f"{base}/{section}/" if section else f"{base}/"


def section_name(section):
    return section.split("/")[-1] if section else "profile"


def is_blocked_url(url):
    return isinstance(url, str) and bool(
        re.search(r"(authwall|/login|/checkpoint/|/uas/login)", url))


def out_dir_for(cwd, name, override=None):
    return str((Path(cwd) / override).resolve()) if override else str(Path(cwd) / f"out-{name}")


def normalize_lines(raw):
    """Trimmed non-empty lines, collapsing runs of the same line."""
    out = []
    for line in str(raw or "").split("\n"):
        t = line.strip()
        if t and (not out or out[-1] != t):
            out.append(t)
    return out


def slice_section(raw, heading, end_markers=SECTION_END_MARKERS):
    """Lines from `heading` (exclusive) to the first end marker (exclusive)."""
    lines = normalize_lines(raw)
    try:
        start = lines.index(heading) + 1
    except ValueError:
        start = 0
    end = len(lines)
    for i in range(start, len(lines)):
        if any(lines[i] == m or lines[i].startswith(m) for m in end_markers):
            end = i
            break
    return lines[start:end]


def split_company_type(line):
    parts = [p.strip() for p in str(line).split("·")]
    return {"company": parts[0] or None,
            "employmentType": parts[1] if len(parts) > 1 else None}


def split_location_arrangement(line):
    parts = [p.strip() for p in str(line).split("·")]
    arr = parts[-1] if len(parts) > 1 else None
    if arr and re.fullmatch(r"(Remote|Hybrid|On-site|Onsite)", arr, re.I):
        return {"location": " · ".join(parts[:-1]) or None, "arrangement": arr}
    return {"location": line.strip() or None, "arrangement": None}


DATE_LINE = re.compile(r"(present|\d{4}|jan|feb|mar|apr|may|jun|jul|aug|sep|oct|nov|dec)", re.I)
DURATION_TAIL = re.compile(r"·\s*[^·]*\b(yr|yrs|mo|mos)\b[^·]*$", re.I)


def looks_like_date_line(line):
    return bool(DATE_LINE.search(line)) and bool(re.search(r"[-–—]|present", line, re.I))


def strip_duration(line):
    return DURATION_TAIL.sub("", str(line)).strip()


def parse_experience(raw):
    """Best-effort Experience parse; see test fixtures for the shape."""
    lines = slice_section(raw, "Experience")
    positions = []
    i = 0
    while i < len(lines):
        nxt = lines[i + 1] if i + 1 < len(lines) else None
        if nxt and "·" in nxt and not looks_like_date_line(nxt) and not looks_like_date_line(lines[i]):
            ct = split_company_type(nxt)
            pos = {"title": lines[i], "company": ct["company"],
                   "employmentType": ct["employmentType"], "dateRange": None,
                   "location": None, "arrangement": None, "description": []}
            j = i + 2
            if j < len(lines) and looks_like_date_line(lines[j]):
                pos["dateRange"] = strip_duration(lines[j])
                j += 1
            if (j < len(lines) and not looks_like_date_line(lines[j])
                    and not (j + 1 < len(lines) and "·" in lines[j + 1]
                             and not looks_like_date_line(lines[j + 1]))):
                la = split_location_arrangement(lines[j])
                pos["location"], pos["arrangement"] = la["location"], la["arrangement"]
                j += 1
            while j < len(lines):
                l = lines[j]
                peek = lines[j + 1] if j + 1 < len(lines) else None
                if peek and "·" in peek and not looks_like_date_line(peek):
                    break  # next entry
                if (re.search(r"\band \+\d+ skills?$", l) or re.fullmatch(r"…?\s*more", l, re.I)
                        or l == "Show translation"):
                    j += 1
                    continue
                pos["description"].append(l)
                j += 1
            positions.append(pos)
            i = j
        else:
            i += 1
    return positions


def parse_education(raw):
    """Best-effort Education parse: school / 'Degree, Field' / date range."""
    lines = slice_section(raw, "Education")
    edu = []
    i = 0
    while i < len(lines):
        detail = lines[i + 1] if i + 1 < len(lines) else None
        date = lines[i + 2] if i + 2 < len(lines) else None
        if (detail and date and looks_like_date_line(date)
                and not looks_like_date_line(lines[i]) and not looks_like_date_line(detail)):
            degree, _, rest = detail.partition(",")
            edu.append({
                "school": lines[i],
                "degree": degree.strip() if rest else None,
                "field": rest.strip() if rest else degree.strip(),
                "dateRange": strip_duration(date),
            })
            i += 3
        else:
            i += 1
    return edu


def parse_skills(raw):
    """Best-effort Skills parse -> [{name, endorsements}]."""
    lines = [l for l in slice_section(raw, "Skills") if l != "All"]
    skills = []

    def is_noise(l):
        return (re.search(r"\bendorsement", l, re.I) or l == "Endorsed"
                or " at " in l or l.startswith("Show"))

    for i, l in enumerate(lines):
        if is_noise(l):
            continue
        endorsements = 0
        for k in range(1, 4):
            if i + k < len(lines):
                m = re.search(r"(\d+)\s+endorsement", lines[i + k], re.I)
                if m:
                    endorsements = int(m.group(1))
                    break
        if not any(s["name"] == l for s in skills):
            skills.append({"name": l, "endorsements": endorsements})
    return skills


def extract_email(text):
    m = re.search(r"[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,}", str(text or ""), re.I)
    return m.group(0) if m else None


def parse_job_id(value):
    """Numeric job id from a /jobs/view/ URL or a bare id."""
    if not isinstance(value, str):
        return None
    s = value.strip()
    s = re.split(r"[?#]", s)[0]
    if re.fullmatch(r"\d{6,}", s):
        return s
    m = re.search(r"/jobs/view/(\d+)", s)
    return m.group(1) if m else None


def job_url(job_id):
    return f"https://www.linkedin.com/jobs/view/{job_id}/"


def parse_job_text(raw):
    """Best-effort job-page parse.

    Anchors on the meta line ('Location · N days ago · M applicants'): the line
    before it is the title, the one before that the company. Description =
    lines after 'About the job' up to the first end marker.
    """
    lines = normalize_lines(raw)
    meta_idx = next((i for i, l in enumerate(lines)
                     if "·" in l and re.search(r"\bapplicants?\b|\bago\b", l, re.I)), -1)
    title = lines[meta_idx - 1] if meta_idx > 0 else None
    company = lines[meta_idx - 2] if meta_idx > 1 else None
    location = (lines[meta_idx].split("·")[0].strip() or None) if meta_idx >= 0 else None
    description = []
    try:
        start = lines.index("About the job") + 1
        end = len(lines)
        for i in range(start, len(lines)):
            if any(lines[i] == m or lines[i].startswith(m) for m in JOB_END_MARKERS):
                end = i
                break
        description = lines[start:end]
    except ValueError:
        pass
    return {"title": title, "company": company, "location": location,
            "description": description}


# ---------------------------------------------------------------------------
# Browser runner (only under __main__)
# ---------------------------------------------------------------------------

def _connect():
    from playwright.sync_api import sync_playwright
    p = sync_playwright().start()
    try:
        browser = p.chromium.connect_over_cdp(CDP_ENDPOINT)
    except Exception:
        print(f"Cannot reach Chrome at {CDP_ENDPOINT}. Start it and log into "
              "LinkedIn once (see header).", file=sys.stderr)
        raise SystemExit(1)
    ctx = browser.contexts[0]
    page = next((pg for pg in ctx.pages if "linkedin.com" in pg.url), None) or ctx.new_page()
    page.set_viewport_size({"width": 1440, "height": 1000})
    return p, browser, page


EXPAND_JS = """async () => {
  const a = document.querySelector('aside.msg-overlay-list-bubble, aside#msg-overlay');
  if (a) a.style.display = 'none';
  document.querySelectorAll('main button').forEach((b) => {
    if ((b.innerText || '').trim().toLowerCase().includes('see more')) b.click();
  });
  const s = document.scrollingElement;
  for (let y = 0; y < s.scrollHeight; y += 700) {
    s.scrollTop = y; await new Promise((r) => setTimeout(r, 150));
  }
  s.scrollTop = 0;
}"""

MAIN_TEXT_JS = "() => (document.querySelector('main') || document.body).innerText"


def _grab(page, url):
    page.goto(url, wait_until="domcontentloaded", timeout=60000)
    time.sleep(3.5)
    if is_blocked_url(page.url):
        print("\nNOT LOGGED IN - log into LinkedIn in the Chrome window first.",
              file=sys.stderr)
        raise SystemExit(2)
    page.evaluate(EXPAND_JS)
    time.sleep(0.8)
    return page.evaluate(MAIN_TEXT_JS)


def run_profile(public_id, out_dir):
    raw_dir = Path(out_dir) / "raw"
    raw_dir.mkdir(parents=True, exist_ok=True)
    p, browser, page = _connect()
    texts = {}
    for section in SECTIONS:
        name = section_name(section)
        print(f"  {name} … ", end="", file=sys.stderr, flush=True)
        txt = _grab(page, section_url(public_id, section))
        texts[name] = txt
        (raw_dir / f"{name}.txt").write_text(txt, encoding="utf-8")
        try:
            page.screenshot(path=str(raw_dir / f"{name}.png"), full_page=True)
        except Exception:
            pass
        print(f"{len(txt)} chars", file=sys.stderr)

    # Contact-info modal (email lives here, not on the section pages).
    contact_email = None
    try:
        print("  contact … ", end="", file=sys.stderr, flush=True)
        page.goto(profile_base(public_id) + "/", wait_until="domcontentloaded", timeout=60000)
        time.sleep(3)
        page.evaluate("""() => {
          for (const el of document.querySelectorAll('a,button,span')) {
            const own = [...el.childNodes].filter((n) => n.nodeType === 3)
              .map((n) => n.textContent.trim()).join('');
            if (own === 'Contact info') { el.click(); return; }
          }
        }""")
        time.sleep(2)
        modal = page.evaluate("""() => {
          const m = document.querySelector('[role="dialog"], .artdeco-modal');
          return m ? m.innerText : null;
        }""")
        contact_email = extract_email(modal)
        if modal:
            (raw_dir / "contact.txt").write_text(modal, encoding="utf-8")
        page.keyboard.press("Escape")
        print(contact_email or "(none)", file=sys.stderr)
    except Exception as e:
        print(f"(skipped: {e})", file=sys.stderr)

    (Path(out_dir) / "profile.raw.txt").write_text(
        "\n".join(f"\n===== {k} =====\n{v}" for k, v in texts.items()), encoding="utf-8")
    structured = {
        "schema": SCHEMA_PROFILE,
        "publicId": public_id,
        "source": profile_base(public_id),
        "contact": {"email": contact_email},
        "positions": parse_experience(texts.get("experience", "")),
        "education": parse_education(texts.get("education", "")),
        "skills": parse_skills(texts.get("skills", "")),
    }
    (Path(out_dir) / "profile.json").write_text(
        json.dumps(structured, indent=2, ensure_ascii=False), encoding="utf-8")

    print(f"\nDone -> {out_dir}", file=sys.stderr)
    print("  raw/*.txt + raw/*.png per section, contact.txt", file=sys.stderr)
    print("  profile.json      (best-effort structured parse)", file=sys.stderr)
    print("  profile.raw.txt   (feed to an LLM for a rich parse)", file=sys.stderr)
    browser.close()
    p.stop()


def run_job(job_id, out_dir):
    raw_dir = Path(out_dir) / "raw"
    raw_dir.mkdir(parents=True, exist_ok=True)
    p, browser, page = _connect()
    print("  job … ", end="", file=sys.stderr, flush=True)
    txt = _grab(page, job_url(job_id))
    (raw_dir / "job.txt").write_text(txt, encoding="utf-8")
    try:
        page.screenshot(path=str(raw_dir / "job.png"), full_page=True)
    except Exception:
        pass
    print(f"{len(txt)} chars", file=sys.stderr)
    parsed = parse_job_text(txt)
    job = {"schema": SCHEMA_JOB, "jobId": job_id, "source": job_url(job_id), **parsed}
    (Path(out_dir) / "job.json").write_text(
        json.dumps(job, indent=2, ensure_ascii=False), encoding="utf-8")
    print(f"\nDone -> {out_dir}", file=sys.stderr)
    print("  raw/job.txt + raw/job.png, job.json", file=sys.stderr)
    browser.close()
    p.stop()


def main(argv):
    if not argv:
        print("usage: uv run li_scrape.py <profile-url-or-id | jobs-view-url-or-job-id> [outDir]",
              file=sys.stderr)
        return 1
    target, override = argv[0], (argv[1] if len(argv) > 1 else None)
    job_id = parse_job_id(target) if ("/jobs/" in target or re.fullmatch(r"\d{6,}", target.strip())) else None
    if job_id:
        run_job(job_id, out_dir_for(Path.cwd(), f"job-{job_id}", override))
        return 0
    public_id = parse_public_id(target)
    if not public_id:
        print(f"could not parse a public id or job id from: {target}", file=sys.stderr)
        return 1
    run_profile(public_id, out_dir_for(Path.cwd(), public_id, override))
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))
